File: src/GamePanel.java
import javax.swing.JPanel;
import java.awt.*;

public class GamePanel extends JPanel implements Runnable{

    //RENDER SETTINGS
    final int originalTileSize =  20;
    final int scale = 2;
    final int tileSize = originalTileSize * scale;
    final int maxScreenCol = 10;
    final int maxScreenRow= 10;
    final int screenWidth = maxScreenCol * tileSize;
    final int screenHeight = maxScreenRow * tileSize;

    int FPS = 60;

    int playerX = 0;
    int playerY = 0;

    int playerSpeed = 2;

    Thread gameThread;
    KeyHandler keyHandler = new KeyHandler();

    //GamePanel constructor for applying the window settings
    public GamePanel(){
        this.setPreferredSize(new Dimension(screenWidth, screenHeight));
        this.setBackground(Color.red);
        this.setDoubleBuffered(true);
        this.addKeyListener(keyHandler);
        this.setFocusable(true);
    }

    public void startGameThread(){
        gameThread = new Thread(this);
       gameThread.start();
    }
    @Override
    public void run() {

        double drawInterval = (double) 1000000000 /FPS;
        double nextDrawTime = System.nanoTime() + drawInterval;
        while (gameThread != null) {
            //System.out.println("Game Thread is active! LESGOO");
            //1. Update
            update();
            //2. DrawCall
            repaint();


            try {
                double remainingTime = nextDrawTime - System.nanoTime();
                remainingTime = remainingTime/1000000;
                if(remainingTime < 0){
                    remainingTime = 0;
                }
                Thread.sleep((long) remainingTime);

                nextDrawTime += drawInterval;
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public void update(){
        if(keyHandler.downPressed){
            playerY += playerSpeed;
        }
        if(keyHandler.upPressed){
            playerY -= playerSpeed;
        }
        if(keyHandler.leftPressed){
            playerX -= playerSpeed;
        }
        if(keyHandler.rightPressed){
            playerX += playerSpeed;
        }
    }

    public void paintComponent(Graphics graphics){

        super.paintComponent(graphics);

        Graphics2D graphics2D = (Graphics2D) graphics;
        graphics2D.setColor(Color.WHITE);
        graphics2D.fillRect(playerX, playerY, tileSize, tileSize);

    }
}

File: src/GamePanel.kt
import java.awt.*
import javax.swing.JPanel

class GamePanel : JPanel(), Runnable {
    // RENDER SETTINGS
    val screenWidth = 1000
    val screenHeight = 800
    var FPS = 60

    var gameThread: Thread? = null
    val objectHandler = ObjectHandler()
    var keyHandler = KeyHandler()
    var physics = Physics(objectHandler)


    fun initialize(){
        val obj1 = GameObject(color = Color	(219,100,100), keyHandler = keyHandler, speed = 6)
        val obj2 = GameObject(color = Color(46,117,130), keyHandler = keyHandler)
        val obj3 = GameObject(color = Color	(109,204,109), keyHandler = keyHandler, speed = 4)

        objectHandler.addObject(obj1)
        objectHandler.addObject(obj2)
        objectHandler.addObject(obj3)

        val cmd = cmd(objectHandler)
        cmd.start()
    }
    // GamePanel constructor for applying the window settings
    init {
        // Set JPanel properties
        this.preferredSize = Dimension(screenWidth, screenHeight) // Size
        setBackground(Color.black) // Background color
        this.isDoubleBuffered = true // Double buffering
        addKeyListener(keyHandler) // Add KeyListener
        setFocusable(true) // Set focusable
    }

    // Method to get the KeyHandler instance
    fun sendKeyHandler(): KeyHandler {
        return keyHandler
    }

    // Method to start the game thread
    fun startGameThread() {
        gameThread = Thread(this)
        gameThread!!.start()

    }

    // Main game loop
    override fun run() {

        var previousTime = System.nanoTime()
        // Timing variables for consistent frame rate
        val drawInterval = 1000000000.0 / FPS
        var nextDrawTime = System.nanoTime() + drawInterval

        // Game loop
        while (gameThread != null) {
            val currentTime = System.nanoTime() // Get the current time
            val deltaTime = (currentTime - previousTime) / 1_000_000_000.0 // Calculate delta time in seconds
            previousTime = currentTime

            // Update game state
            update(deltaTime)
            // Repaint the screen
            repaint()
            // Control frame rate
            try {
                var remainingTime = nextDrawTime - System.nanoTime()
                remainingTime = remainingTime / 1000000
                if (remainingTime < 0) {
                    remainingTime = 0.0
                }
                Thread.sleep(remainingTime.toLong())
                nextDrawTime += drawInterval
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
        }
    }

    // Method to update game state
    fun update(deltaTime: Double) {
        objectHandler.updateAll()
        physics.applyGravity()
    }

    // Method to render graphics
    public override fun paintComponent(graphics: Graphics) {
        super.paintComponent(graphics)
        val graphics2D = graphics as Graphics2D
        objectHandler.renderAll(graphics2D)
        graphics2D.font = Font("Arial", Font.BOLD, 30) // Change font size
        graphics2D.color = Color.white // Change text color
        graphics2D.drawString("Send help", 1370, 30)
    }
}